test/test.ads.js

```diff
@@ -889,85 +684,115 @@ QUnit.test('startLinearAdMode should only trigger adstart from correct states',
+
+  this.player.ads.endLinearAdMode();
+  this.player.trigger('playing');
+
   this.player.ads.startLinearAdMode();
-  this.player.ads.state = 'ads-ready';
+  assert.strictEqual(adstart.callCount, 2, 'Midroll');
+
   this.player.ads.startLinearAdMode();
-  this.player.ads.state = 'ad-playback';
+  assert.strictEqual(adstart.callCount, 2, 'During midroll playback');
+
+  this.player.ads.endLinearAdMode();
+  this.player.trigger('playing');
+
+  this.player.trigger('ended');
+  this.player.ads.startLinearAdMode();
+  assert.strictEqual(adstart.callCount, 3, 'Postroll');
+
   this.player.ads.startLinearAdMode();
-  assert.strictEqual(adstart.callCount, 3, 'other states');
+  assert.strictEqual(adstart.callCount, 3, 'During postroll playback');
+
+  this.player.ads.endLinearAdMode();
+  assert.strictEqual(adstart.callCount, 3, 'Ads done');
+  
 });
 
 QUnit.test('ad impl can notify contrib-ads there is no preroll', function(assert) {
 
-  this.player.ads.state = 'preroll?';
+  this.player.trigger('loadstart');
+  this.player.trigger('nopreroll');
+  this.player.trigger('play');
+  this.player.trigger('adsready');
+
+  assert.strictEqual(this.player.ads.isInAdMode(), false, 'not in ad mode');
+
+});
+
+// Same test as above with different event order because this used to be broken.
+QUnit.test('ad impl can notify contrib-ads there is no preroll (adsready before play)', function(assert) {
+
+  this.player.trigger('loadstart');
   this.player.trigger('nopreroll');
-  assert.strictEqual(this.player.ads.state, 'content-playback', 'no longer in preroll?');
+  this.player.trigger('adsready');
+  this.player.trigger('play');
+
+  assert.strictEqual(this.player.ads.isInAdMode(), false, 'not in ad mode');
 
 });
 
 QUnit.test('ad impl can notify contrib-ads there is no postroll', function(assert) {
 
+  var ended = sinon.spy();
+
+  this.player.on('ended', ended);
+
+  this.player.trigger('loadstart');
   this.player.trigger('nopostroll');
-  this.player.ads.state = 'content-playback';
+  this.player.trigger('play');
+  this.player.trigger('adsready');
+  this.player.ads.skipLinearAdMode();
   this.player.trigger('contentended');
-  this.clock.tick(5);
-  assert.strictEqual(this.player.ads.state, 'content-playback', 'no longer in postroll?');
+  this.clock.tick(1);
+  assert.ok(ended.calledOnce, 'Ended triggered');
 
 });
 
 QUnit.test('ended event is sent with postroll', function(assert) {
 
-  var ended = sinon.spy();
+  this.player.trigger('loadstart');
+  this.player.trigger('adsready');
+  this.player.trigger('play');
+  this.player.ads.skipLinearAdMode();
 
-  this.player.tech_.el_ = {
-    ended: true,
-    hasChildNodes: function() {
-      return false;
-    },
-    removeAttribute: function() {
+  var ended = sinon.spy();
 
-    }
-  };
   this.player.on('ended', ended);
-  this.player.ads.state = 'content-playback';
+
   this.player.trigger('contentended');
 
   this.clock.tick(10000);
-
   assert.ok(ended.calledOnce, 'Ended triggered');
```

Comment:
seems like this is a duplicate of the previous test now

Reply:
removing the duplicate test